### ISSUE TITLE
All props are passed to DOM elements in React 16

### ISSUE THREAD

**Issue by faddee** [rtsao/styletron#167]
This test #166 shows that the hack that disables the unknown prop warning in React 15.2.0 is [removed in version 16](https://github.com/facebook/react/pull/6459) and all props are passed to DOM elements: https://facebook.github.io/react/blog/2017/09/08/dom-attributes-in-react-16.html

@rtsao how do we approach this?

What I can think of
1. EDIT ~~Revert commit b418d600e1c637b796cfd2b5c6788e6bc00965d6 and whitelist props again~~ [not recommended](https://twitter.com/dan_abramov/status/822526388557250560)
1. Provide a third argument where users can omit props
    ```js
    styled(
      'div',
      ({isChristmas}) => ({backound: isChristmas ? 'red' : 'grey'}),
      ({isChristmas, ...rest}) => rest
    );
    ```

I personally don't like any of these be cause one of them adds white listing that was removed in React for a reason and the other one forces the user perform omitting manually.

**Comment by faddee** [comment]
The third option I have and maybe the most straight forward is to create a HOC which deletes unwanted props before you pass them.
```js
styled(
  ({isChristmas, ...rest}) => <div {...rest} />,
  ({isChristmas}) => ({backound: isChristmas ? 'red' : 'grey'})
);
```

But when composing other style components with this, you may encounter unwanted behavior.

**Comment by rtsao** [comment]
Yeah, this whole situation is rather unfortunate.

A while back I built a Babel plugin to statically analyze which props were used then blacklist them for you (via an extra argument). I think this would actually work for most use cases, but seems a bit heavy-handed in terms of magic. I was also concerned about robustness, but I think it would actually be pretty good since we could error out and notify the user that props were being consumed in a manner beyond the reach of static analysis. In this case, the user would have to manually write the prop blacklist.

Perhaps this could alleviate the user burden of using your second proposal.

Another option is to simply require style props to be nested under a special prop, but this makes consumption of styled components in JSX a lot worse IMHO.

**Comment by faddee** [comment]
We're using TypeScript where you can't use things like plugins. 😞 This would force us to add an extra compiler to the build chain which won't happen since we already have bad performance. 

Another option I have in mind is by using getters or `Proxy` to determine which props are being used in the style function. But it comes with some nasty overhead by wrapping all props with getters on each render and with `Proxy` we'll have performance and compability problems. It would be easy to try it out though.

I also manage to solve the composition problem in the option I mentioned in my second comment. Instead of passing `className` as a string with pregenerated class names, I passed `className` as a function and merged the style result from each style component it passed when `className.toString()` was called. It was super crazy. I even deleted the branch how crazy it was. 😂 But it actually worked. And support of nested composition is actually something people of mine that are using Styletron are asking for. It's quite often we need to wrap styled components with HOCs to the the typing right.

**Comment by faddee** [comment]
A specific prop called `styledProp` will be deleted before passing to the inner component. Fixed in #169
